File: tests/e2e/test_zk_rebalancing_workflow.py
#!/usr/bin/env python3
"""
ZK Rebalancing End-to-End Test

Tests the complete workflow of zero-knowledge proof based portfolio rebalancing
with ERC-8004 agentic orchestration.
"""

import os
import sys
import time
from web3 import Web3

# Add parent directories to path for imports
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../..'))

from agents.rebalancer_agent import RebalancerAgent
from agents.validator_agent import ValidatorAgent
from agents.client_agent import ClientAgent


def test_zk_rebalancing_e2e():
    """Test complete end-to-end ZK rebalancing workflow"""
    print("\n" + "="*70)
    print("  ZK Rebalancing Proof - End-to-End Test")
    print("  ERC-8004 Agentic Orchestration")
    print("="*70 + "\n")

    # Connect to blockchain
    w3 = Web3(Web3.HTTPProvider(os.getenv("RPC_URL", "http://localhost:8545")))
    assert w3.is_connected(), "❌ Blockchain not connected"
    print("✅ Connected to blockchain")
    
    # Create unique agents
    timestamp = int(time.time() * 1000)
    
    # Initialize agents
    print("\n" + "─"*70)
    print("STEP 1: Initialize Agents")
    print("─"*70)
    
    rebalancer = RebalancerAgent(
        agent_domain=f"rebalancer-{timestamp}.zk-proof.test",
        private_key="0x" + os.urandom(32).hex()
    )
    print(f"   Rebalancer: {rebalancer.address}")
    
    validator = ValidatorAgent(
        agent_domain=f"validator-{timestamp}.zk-proof.test",
        private_key="0x" + os.urandom(32).hex()
    )
    print(f"   Validator: {validator.address}")
    
    client = ClientAgent(
        agent_domain=f"client-{timestamp}.zk-proof.test",
        private_key="0x" + os.urandom(32).hex()
    )
    print(f"   Client: {client.address}")
    
    # Fund all agents
    print("\n" + "─"*70)
    print("STEP 2: Fund Agents")
    print("─"*70)
    
    for agent in [rebalancer, validator, client]:
        if w3.eth.get_balance(agent.address) < w3.to_wei(0.1, 'ether'):
            tx = {
                'from': w3.eth.accounts[0],
                'to': agent.address,
                'value': w3.to_wei(0.5, 'ether'),
                'gas': 21000,
                'gasPrice': w3.eth.gas_price
            }
            tx_hash = w3.eth.send_transaction(tx)
            w3.eth.wait_for_transaction_receipt(tx_hash)
    print("✅ All agents funded with 0.5 ETH")
    
    # Register agents
    print("\n" + "─"*70)
    print("STEP 3: Register Agents on ERC-8004 Registry")
    print("─"*70)
    
    rebalancer.register_agent()
    print(f"   Rebalancer Agent ID: {rebalancer.agent_id}")
    
    validator.register_agent()
    print(f"   Validator Agent ID: {validator.agent_id}")
    
    client.register_agent()
    print(f"   Client Agent ID: {client.agent_id}")
    
    # Create rebalancing plan
    print("\n" + "─"*70)
    print("STEP 4: Create Rebalancing Plan")
    print("─"*70)
    
    # Example: 4-asset portfolio rebalancing
    # Old: 1000, 1000, 1000, 750 tokens at price 100 each
    # New: 800, 800, 1200, 950 tokens (same total value)
    rebalancing_plan = rebalancer.create_rebalancing_plan(
        old_balances=["1000", "1000", "1000", "750"],
        new_balances=["800", "800", "1200", "950"],
        prices=["100", "100", "100", "100"],
        min_allocation_pct="10",
        max_allocation_pct="40"
    )
    print(f"   Portfolio value: {rebalancing_plan['new_total_value']:,}")
    
    # Generate ZK proof
    print("\n" + "─"*70)
    print("STEP 5: Generate Zero-Knowledge Proof")
    print("─"*70)
    
    proof_package = rebalancer.generate_zk_proof(rebalancing_plan)
    print(f"   Proof generated using {proof_package['metadata']['proof_system']}")
    print(f"   Curve: {proof_package['metadata']['curve']}")
    print(f"   Public inputs: {len(proof_package['public_inputs'])} signals")
    
    # Submit proof for validation
    print("\n" + "─"*70)
    print("STEP 6: Submit Proof for Validation")
    print("─"*70)
    
    validation_tx = rebalancer.submit_proof_for_validation(
        proof_package,
        validator.agent_id
    )
    print(f"   Validation request transaction: {validation_tx}")
    
    # Validator validates the proof
    print("\n" + "─"*70)
    print("STEP 7: Validate ZK Proof")
    print("─"*70)
    
    validation_result = validator.validate_proof(proof_package)
    print(f"   Proof valid: {validation_result['is_valid']}")
    print(f"   Validation score: {validation_result['score']}/100")
    
    # Submit validation response
    print("\n" + "─"*70)
    print("STEP 8: Submit Validation Response")
    print("─"*70)
    
    validation_response_tx = validator.submit_validation_response(
        validation_result['validation_package']
    )
    print(f"   Validation response transaction: {validation_response_tx}")
    
    # Authorize client feedback
    print("\n" + "─"*70)
    print("STEP 9: Authorize Client Feedback")
    print("─"*70)
    
    auth_tx = rebalancer.authorize_client_feedback(client.agent_id)
    print(f"   Authorization transaction: {auth_tx}")
    
    # Client evaluates and provides feedback
    print("\n" + "─"*70)
    print("STEP 10: Client Evaluation and Feedback")
    print("─"*70)
    
    quality_score = client.evaluate_rebalancing_quality(proof_package)
    print(f"   Quality evaluation: {quality_score}/100")
    
    feedback = client.submit_feedback(
        server_agent_id=rebalancer.agent_id,
        score=quality_score,
        comment="Excellent ZK proof-based rebalancing service with strong privacy guarantees"
    )
    print(f"   Feedback submitted: {feedback['score']}/100")
    
    # Check reputation
    print("\n" + "─"*70)
    print("STEP 11: Check Rebalancer Reputation")
    print("─"*70)
    
    reputation = client.check_rebalancer_reputation(rebalancer.agent_id)
    print(f"   Feedback count: {reputation['feedback_count']}")
    print(f"   Average score: {reputation['average_score']:.1f}/100")
    
    # Summary
    print("\n" + "="*70)
    print("  ✅ END-TO-END TEST COMPLETE")
    print("="*70)
    print("\nWorkflow Summary:")
    print("  1. ✅ Three agents initialized (Rebalancer, Validator, Client)")
    print("  2. ✅ Agents registered on ERC-8004 registry")
    print("  3. ✅ Rebalancing plan created (4-asset portfolio)")
    print("  4. ✅ Zero-knowledge proof generated (Groth16)")
    print("  5. ✅ Proof validated cryptographically")
    print("  6. ✅ Validation response submitted on-chain")
    print("  7. ✅ Client feedback authorized and submitted")
    print("  8. ✅ Reputation tracking operational")
    print("\nKey Benefits Demonstrated:")
    print("  • Privacy: Portfolio positions hidden via ZK proofs")
    print("  • Trust: Cryptographic validation of rebalancing constraints")
    print("  • Transparency: All interactions recorded on-chain")
    print("  • Reputation: Feedback system for service quality")
    print("\n" + "="*70 + "\n")
    
    return True


if __name__ == "__main__":
    try:
        success = test_zk_rebalancing_e2e()
        sys.exit(0 if success else 1)
    except Exception as e:
        print(f"\n❌ Test failed: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

File: tests/e2e/test-zk-rebalancing-workflow.ts
#!/usr/bin/env ts-node

/**
 * ZK Rebalancer Validation E2E Test
 * Tests the complete workflow using ZyFI's rebalancer validation rules
 */

import { createPublicClient, createWalletClient, http, parseEther } from "viem";
import { privateKeyToAccount } from "viem/accounts";
import { foundry } from "viem/chains";
import { randomBytes } from "crypto";
import { execSync } from "child_process";
import { readFileSync } from "fs";
import { join } from "path";
import { RebalancerAgent, ValidatorAgent, ClientAgent } from "../../agents";

async function deployContracts(): Promise<void> {
  console.log("─".repeat(70));
  console.log("STEP 0: Deploy Contracts");
  console.log("─".repeat(70));

  try {
    // Note: Run `npm run setup:zkp` manually once when circuit changes.
    // For normal E2E testing with different inputs, setup is NOT needed.
    // The RebalancerVerifier.sol and Verifier.sol remain constant for the same rebalancer validation circuit.

    execSync("npm run forge:deploy:local", {
      stdio: "inherit",
      cwd: process.cwd(),
    });
    console.log("Contracts deployed\n");
  } catch (error) {
    throw new Error("Failed to deploy contracts. Is Anvil running?");
  }
}

async function testZkRebalancingE2E(): Promise<void> {
  console.log("\n" + "=".repeat(70));
  console.log("  ZK Rebalancer Validation - End-to-End Test");
  console.log("  Using ZyFI Backend Validation Rules");
  console.log("=".repeat(70) + "\n");

  const rpcUrl = process.env.RPC_URL || "http://127.0.0.1:8545";
  const publicClient = createPublicClient({
    chain: foundry,
    transport: http(rpcUrl),
  });

  console.log("Connected to blockchain\n");

  // Deploy contracts
  await deployContracts();

  const timestamp = Date.now();

  // Initialize agents
  console.log("─".repeat(70));
  console.log("STEP 1: Initialize Agents");
  console.log("─".repeat(70));

  const rebalancer = new RebalancerAgent(
    `rebalancer-${timestamp}.test`,
    `0x${randomBytes(32).toString("hex")}` as `0x${string}`
  );

  const validator = new ValidatorAgent(
    `validator-${timestamp}.test`,
    `0x${randomBytes(32).toString("hex")}` as `0x${string}`
  );

  const client = new ClientAgent(
    `client-${timestamp}.test`,
    `0x${randomBytes(32).toString("hex")}` as `0x${string}`
  );

  // Fund agents
  console.log("\n" + "─".repeat(70));
  console.log("STEP 2: Fund Agents");
  console.log("─".repeat(70));

  const fundingAccount = privateKeyToAccount(
    "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80"
  );

  const walletClient = createWalletClient({
    account: fundingAccount,
    chain: foundry,
    transport: http(rpcUrl),
  });

  for (const agent of [rebalancer, validator, client]) {
    const balance = await publicClient.getBalance({ address: agent.address });
    if (balance < parseEther("0.1")) {
      const hash = await walletClient.sendTransaction({
        to: agent.address,
        value: parseEther("0.5"),
      });
      await publicClient.waitForTransactionReceipt({ hash });
    }
  }

  console.log("All agents funded\n");

  // Register agents
  console.log("─".repeat(70));
  console.log("STEP 3: Register Agents");
  console.log("─".repeat(70));

  await rebalancer.registerAgent();
  await validator.registerAgent();
  await client.registerAgent();

  // Load rebalancer validation input data
  console.log("\n" + "─".repeat(70));
  console.log("STEP 4: Load Rebalancer Validation Input");
  console.log("─".repeat(70));

  const rebalancerInputPath = join(
    process.cwd(),
    "input",
    "rebalancer-input.json"
  );
  const rebalancerData = JSON.parse(readFileSync(rebalancerInputPath, "utf-8"));

  console.log(`📂 Loaded from: input/rebalancer-input.json`);
  console.log(`   Liquidity: $${rebalancerData.liquidity.toLocaleString()}`);
  console.log(`   ZyFI TVL: $${rebalancerData.zyfiTvl.toLocaleString()}`);
  console.log(
    `   Rebalancer Amount: ${rebalancerData.amount.toLocaleString()}`
  );
  console.log(`   Pool TVL: ${rebalancerData.poolTvl.toLocaleString()}`);
  console.log(`   New APY: ${rebalancerData.newApy / 10000}%`);
  console.log(`   Old APY: ${rebalancerData.oldApy / 10000}%`);

  // Generate rebalancer validation proof
  console.log("\n" + "─".repeat(70));
  console.log("STEP 5: Generate Rebalancer Validation ZK Proof");
  console.log("─".repeat(70));

  const proof = rebalancer.generateRebalancerValidationProof({
    // New opportunity data
    liquidity: rebalancerData.liquidity,
    zyfiTvl: rebalancerData.zyfiTvl,
    amount: rebalancerData.amount,
    poolTvl: rebalancerData.poolTvl,
    newApy: rebalancerData.newApy,
    apyStable7Days: rebalancerData.apyStable7Days,
    tvlStable: rebalancerData.tvlStable,
    // Old opportunity data
    oldApy: rebalancerData.oldApy,
    oldLiquidity: rebalancerData.oldLiquidity ?? 0,
    oldZyfiTvl: rebalancerData.oldZyfiTvl ?? 0,
    oldTvlStable: rebalancerData.oldTvlStable ?? 1,
    oldUtilizationStable: rebalancerData.oldUtilizationStable ?? 1,
    oldCollateralHealth: rebalancerData.oldCollateralHealth ?? 1,
    oldZyfiTVLCheck: rebalancerData.oldZyfiTVLCheck ?? 1,
    // User preferences
    supportsCurrentPool: rebalancerData.supportsCurrentPool ?? 1,
  });

  // Submit for validation
  console.log("\n" + "─".repeat(70));
  console.log("STEP 6: Submit for Validation");
  console.log("─".repeat(70));

  await rebalancer.requestValidationFromValidator(proof, validator.address);

  // Validate
  console.log("\n" + "─".repeat(70));
  console.log("STEP 7: Validate Proof");
  console.log("─".repeat(70));

  const validationResult = await validator.validateProof(proof);

  // Submit validation
  console.log("\n" + "─".repeat(70));
  console.log("STEP 8: Submit Validation");
  console.log("─".repeat(70));

  await validator.submitValidation(validationResult);

  // Generate feedback authorization
  console.log("\n" + "─".repeat(70));
  console.log("STEP 9: Generate Feedback Authorization");
  console.log("─".repeat(70));

  const { feedbackAuth } = await rebalancer.generateFeedbackAuthorization(
    client.address,
    10n,
    30
  );

  // Evaluate and feedback
  console.log("\n" + "─".repeat(70));
  console.log("STEP 10: Client Feedback");
  console.log("─".repeat(70));

  const score = client.evaluateRebalancingQuality(proof);
  await client.submitFeedback(
    rebalancer.agentId!,
    score,
    feedbackAuth,
    "Great rebalancer validation service!"
  );

  // Check reputation
  console.log("\n" + "─".repeat(70));
  console.log("STEP 11: Check Reputation");
  console.log("─".repeat(70));

  client.checkRebalancerReputation(rebalancer.agentId!);

  // Summary
  console.log("\n" + "=".repeat(70));
  console.log("  REBALANCER VALIDATION TEST COMPLETE");
  console.log("=".repeat(70));
  console.log("\nAll steps executed successfully!");
  console.log(
    "  • Rebalancer validation input loaded from input/rebalancer-input.json"
  );
  console.log("  • ZK proof generated with ZyFI validation constraints:");
  console.log("    - Liquidity constraint verified");
  console.log("    - TVL constraint verified (max 25% allocation)");
  console.log("    - APY performance improvement verified");
  console.log("    - APY stability verified (7 or 10 days)");
  console.log("    - TVL stability verified");
  console.log("  • Proof validated on-chain");
  console.log("  • Agents registered and coordinated");
  console.log("  • Feedback and reputation tracked");
  console.log("\n" + "=".repeat(70) + "\n");
}

if (require.main === module) {
  testZkRebalancingE2E()
    .then(() => {
      console.log("Test completed successfully");
      process.exit(0);
    })
    .catch((error) => {
      console.error("\nTest failed");
      console.error(error);
      process.exit(1);
    });
}

export { testZkRebalancingE2E };
